lib/src/yaml_edit/editor.dart

```diff
@@ -0,0 +1,651 @@
+/// elements in the path is important, and it should be arranged in order of
+/// calling, with the first element being the first key or index to be called.
+///
+/// To dump the YAML after all the modifications have been completed, simply
+/// call [toString()].
+///
+/// [1]: https://yaml.org/
+@sealed
+class YamlEditor {
+  /// List of [SourceEdit]s that have been applied to [_yaml] since the creation
+  /// of this instance, in chronological order. Intended to be compatible with
+  /// `package:analysis_server`.
+  final List<SourceEdit> _edits = [];
+
+  UnmodifiableListView<SourceEdit> get edits =>
+      UnmodifiableListView<SourceEdit>([..._edits]);
+
+  /// Current YAML string.
+  String _yaml;
+
+  /// Root node of YAML AST.
+  YamlNode _contents;
+
+  /// Stores the list of nodes in [_contents] that are connected by aliases.
+  Set<YamlNode> _aliases = {};
+
+  @override
+  String toString() => _yaml;
+
+  /// Returns the detected indentation step used for this YAML document, or
+  /// defaults to a value of `2` if no indentation step can be detected.
+  ///
+  /// Indentation step is determined by the difference in indentation of the
+  /// first block-styled yaml collection in the second level as compared to the
+  /// top-level elements. In the case where there are multiple possible
+  /// candidates, we choose the candidate closest to the start of [_yaml].
+  int get indentation {
+    if (_indentation == null) {
+      final node = _contents;
+      Iterable<YamlNode> children;
+
+      if (node is YamlMap && node.style == CollectionStyle.BLOCK) {
+        children = node.nodes.values;
+      } else if (node is YamlList && node.style == CollectionStyle.BLOCK) {
+        children = node.nodes;
+      }
+
+      if (children != null) {
+        for (var child in children) {
+          var indent = 0;
+          if (child is YamlList) {
+            indent = getListIndentation(child);
+          } else if (child is YamlMap) {
+            indent = getMapIndentation(child);
+          }
+
+          if (indent != 0) _indentation = indent;
+        }
+      }
+    }
+
+    /// If [_indentation] was not set by the above block, give it a default
+    /// value of 2.
+    _indentation ??= 2;
+
+    return _indentation;
+  }
+
+  int _indentation;
+
+  String get lineEnding {
+    if (_isWindowsEnding == null) {
+      var index = -1;
+      var unixNewlines = 0;
+      var windowsNewlines = 0;
+      while ((index = _yaml.indexOf('\n', index + 1)) != -1) {
+        if (index != 0 && _yaml[index - 1] == '\r') {
+          windowsNewlines++;
+        } else {
+          unixNewlines++;
+        }
+      }
+      _isWindowsEnding = windowsNewlines > unixNewlines;
+    }
+
+    return _isWindowsEnding ? '\r\n' : '\n';
+  }
+
+  bool _isWindowsEnding;
+
+  factory YamlEditor(String yaml) => YamlEditor._(yaml);
+
+  YamlEditor._(this._yaml) {
+    ArgumentError.checkNotNull(_yaml);
+    initialize();
+  }
+
+  void initialize() {
```

Comment:
Could this be private?
Could it be inlined in the constructor?

Reply:
I turned it private!

The reason why this was taken out as a function is because when the string is changed after a modification (in `_performEdit`), the YAML tree is reloaded, and we re-analyze the string for aliases/line endings/indentation steps!